server/src/main/java/org/elasticsearch/script/ScriptService.java

```diff
@@ -219,7 +252,16 @@ boolean compilationLimitsEnabled() {
 
     void registerClusterSettingsListeners(ClusterSettings clusterSettings) {
         clusterSettings.addSettingsUpdateConsumer(SCRIPT_MAX_SIZE_IN_BYTES, this::setMaxSizeInBytes);
-        clusterSettings.addSettingsUpdateConsumer(SCRIPT_MAX_COMPILATIONS_RATE, compiler::setMaxCompilationRate);
+        clusterSettings.addSettingsUpdateConsumer(SCRIPT_GENERAL_MAX_COMPILATIONS_RATE, cacheUpdater::setMaxCompilationRate);
```

Comment:
One problem that seems to make this more complicated is having individual update consumers. There is an existing method we can use that takes a list of settings, and passes all the values when any are updated. I think using this pattern will make the code simpler (less locking and special cases). The pattern I'm imagining is something like an update consumer for the following scenarios:

- general settings -> this changes the general cache, and handles flipping between general and per context
- per context + general settings -> the general settings along with concrete settings for each context. if the general cache is used, this would result in an error if any of the per context settings exist. otherwise the ScriptCache for that context would be looked up. All 3 setting values should then be passed into a single update method on ScriptCache, so it can be rebuilt once whether one or all 3 settings change (or the fallbacks for size/expire changed).

Reply:
Done.